### DISCUSSION TITLE
PageResolver

### DISCUSSION THREAD

**Discussion by matt-goldman** [CommunityToolkit/Maui#415] [New Feature Discussions]
Last year I released a [NuGet package](https://www.nuget.org/packages/Goldie.MauiPlugins.PageResolver/) with some extension methods for `INavigation` that use the built in ServiceProvider to resolve dependencies for pages and push them onto the navigation stack. [The wiki](https://github.com/matt-goldman/Maui.Plugins.PageResolver/wiki) has detailed instructions on how to use it and there's a [video on YouTube](https://www.youtube.com/watch?v=qx8A4zIe9dU&t=326s) breaking it down too.

Recently James Montemagno suggested on Twitter that this package might be a good fit for the Community Toolkit. There's also a [discussion](https://github.com/dotnet/maui/issues/3622) on the .NET MAUI repo suggesting that this approach might be a good fit for .NET MAUI itself.

I feel this approach complements Shell by offering the same automatic resolution of dependencies but using INavigation. It's also useful even with Shell for pushing modal pages.

Keen to hear other people's thoughts on this. Do you think this would be a good fit for the Community Toolkit? I would be happy to port it in if people do. 

There's also an add-on package called AutoReg with a source generator to automatically register Pages, ViewModels and Services with the ServiceProvider based on a naming convention. My gut feeling is that this isn't as good a fit for the Community Toolkit, but again if people feel like it is I would be happy to port it in (it's kind of experimental and a PoC at the moment so would need some work).

PS thanks to @bijington for talking me through this process.

**Comment by bijington** [comment]
I really like having the ability to resolve a view and it's dependency tree if not using `Shell`. I know that some existing applications use functionality that is exposed via `INavigation` which `Shell` does not provide so when/if those applications are migrated over to .NET MAUI then they will need to stick with using `INavigation`.

**Comment by CliffAgius** [comment]
Closed as answered
